src/protocols/secure_channel/PASESession.cpp

```diff
@@ -346,133 +349,200 @@ CHIP_ERROR PASESession::DeriveSecureSession(SecureSession & session, SecureSessi
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    SuccessOrExit(err = tlvReader.Get(hasPBKDFParameters));
+
+    err = SendPBKDFParamResponse(ByteSpan(initiatorRandom), hasPBKDFParameters);
     SuccessOrExit(err);
 
 exit:
 
     if (err != CHIP_NO_ERROR)
     {
-        SendErrorMsg(Spake2pErrorType::kUnexpected);
+        SendStatusReport(kProtocolCodeInvalidParam);
     }
     return err;
 }
 
-CHIP_ERROR PASESession::SendPBKDFParamResponse()
+CHIP_ERROR PASESession::SendPBKDFParamResponse(ByteSpan initiatorRandom, bool initiatorHasPBKDFParams)
 {
-    System::PacketBufferHandle resp;
-    static_assert(CHAR_BIT == 8, "Assuming sizeof() returns octets here and for sizeof(mPoint)");
-    size_t resplen = kPBKDFParamRandomNumberSize + sizeof(uint64_t) + sizeof(uint32_t) + mSaltLength;
+    uint8_t responderRandom[kPBKDFParamRandomNumberSize] = { 0 };
+    ReturnErrorOnFailure(DRBG_get_bytes(responderRandom, kPBKDFParamRandomNumberSize));
 
-    size_t sizeof_point = sizeof(mPoint);
-
-    uint8_t * msg = nullptr;
-
-    resp = System::PacketBufferHandle::New(resplen);
+    size_t data_len = EstimateTLVStructOverhead(
+        kPBKDFParamRandomNumberSize + kPBKDFParamRandomNumberSize + sizeof(uint16_t) + sizeof(uint16_t) + sizeof(uint8_t), 5);
+    System::PacketBufferHandle resp = System::PacketBufferHandle::New(data_len);
     VerifyOrReturnError(!resp.IsNull(), CHIP_ERROR_NO_MEMORY);
 
-    msg = resp->Start();
+    System::PacketBufferTLVWriter tlvWriter;
+    tlvWriter.Init(std::move(resp));
 
-    // Fill in the random value
-    ReturnErrorOnFailure(DRBG_get_bytes(msg, kPBKDFParamRandomNumberSize));
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+    ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(1), initiatorRandom));
+    ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(2), responderRandom, sizeof(responderRandom)));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(3), GetLocalKeyId(), true));
 
-    // Let's construct the rest of the message using BufferWriter
+    if (!initiatorHasPBKDFParams)
     {
-        Encoding::LittleEndian::BufferWriter bbuf(&msg[kPBKDFParamRandomNumberSize], resplen - kPBKDFParamRandomNumberSize);
-        bbuf.Put64(mIterationCount);
-        bbuf.Put32(mSaltLength);
-        bbuf.Put(mSalt, mSaltLength);
-        VerifyOrReturnError(bbuf.Fit(), CHIP_ERROR_NO_MEMORY);
+        TLV::TLVType outerContainer;
+        ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::ContextTag(4), TLV::kTLVType_Structure, outerContainer));
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(1), mIterationCount, true));
+        ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(2), mSalt, mSaltLength));
+        ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainer));
```

Comment:
- outerContainer here should be renamed as pbkdfParamContainer or something so that we don't have two "outer containers" whose type tags are "outerContainerType" and "outerContainer"